### DISCUSSION TITLE
Cookies

### DISCUSSION THREAD

**Discussion by sasgosh** [quasarframework/quasar#9749] [General - Components / Directives / etc] [answered]
Hello, 

Quasar project by itself creates empty local storage, sesstion storage and indexedDB object so browser shows that website is using 3 cookies but it doesn't. 

The fact that the browser shows that website has cookies, implies users tracking and must be stated in the privacy policy. How to state that the site does not track users when the browser says it has cookies?

Most users are not professional enough and will certainly not check the contents of local/session/indexDB storage objects, and they identify the presence of cookies with tracking. It turns out that I need to write a notice that the website uses cookies but does not actually use them. It's so confusing. 

How can it be set that the framework does not use cookies at all?

**Comment by MilosPaunovic** [comment]
Hello @SaleCar,

When you say `creates empty objects` do you mean separate segments in the `Application` tab of devtools, because that is browser specific, not related to `Quasar` itself, especially if there are no values in there.

**Reply by sasgosh** [reply]
In Chrome, when click on icon right to url, it shows: Cookies (2 in use).

**Reply by ontwikkelfabriek** [reply]
@SaleCar 

A default created SPA/PWA (Qv1) with Quasar CLI has **0** cookies in Chrome.  

See for example:
https://ipm-toolbox.nl

**Reply by sasgosh** [reply]
![image](https://user-images.githubusercontent.com/39879027/122784303-a5a50880-d2b2-11eb-8c22-ddff94787858.png)

**Reply by ontwikkelfabriek** [reply]
@SaleCar 

Well ok, but I get different results.


![cookies2](https://user-images.githubusercontent.com/9429959/122785618-d0dc2780-d2b3-11eb-9717-d7a73276aa6d.png)

![cookie](https://user-images.githubusercontent.com/9429959/122785623-d174be00-d2b3-11eb-9df6-6243bd9ff9d5.png)

![cookies3](https://user-images.githubusercontent.com/9429959/122786134-4647f800-d2b4-11eb-84e5-45fda5892be6.png)


I don't think there are actual cookies.

**Reply by sasgosh** [reply]
Yes. That! What is that?! Who opens empty local storage and session storage objects so Chome report cookie use?

**Comment by sasgosh** [comment]
When I put 

`<script src="https://cdn.jsdelivr.net/npm/quasar@2.0.0/dist/quasar.umd.js"></script>`

in standard empty index.html file... Chrome shows Cookies (2 in use)

When remove that line, it shows Cookies (0 in use)

So, Quasar by itself creates empty local storage and session storage objects.

**Comment by ontwikkelfabriek** [comment]
@rstoenescu 

Any idea? ( of course you do ). Please enlighten us!

**Comment by ontwikkelfabriek** [comment]
@SaleCar 

Nobody seems to be responding. I would create a github issue for this. Qv2 gives 6 empty cookies and Qv1 apps 2 empty cookies, that does not seem right at all.

**Reply by sasgosh** [reply]
Thanks, that would be nice. This is not a problem for the end of the world, but we need to see what it is about. I don't see any particular reason for these objects if they are not in use.

**Comment by rstoenescu** [comment, marked as answer]
1. Quasar DOES NOT sets any cookies by its own (!!).
2. Unless specified, Quasar UMD uses all Quasar plugins, which includes the SessionStorage and LocalStorage -- but it just prepares them, it not uses them on its own (!!)
3. It seems that Chrome decided to say "Cookies" even if not cookies are actually reported. To check the cookies, use the Quasar Cookies plugin or use browser devtools:
![cookies](https://user-images.githubusercontent.com/4665641/123171561-a7262a80-d484-11eb-9b0c-5ed99c8c45ea.png)
4. If you're not setting cookies (yourself or by some underlying package) then you don't need to touch any privacy policy. Which is your case from what I gather. No cookies are set on their own, not by Vue or Quasar. Just because Chrome decided to incorrectly label other things as "Cookies" not mean that you are actually using cookies.
5. It would be polite to wait more than 2 days for someone to reply. This is opensource, not a paid customer service. Sorry for ranting here, but I need to yell it.

**Reply by sasgosh** [reply]
You said everything we already know. It’s not a problem whether Quasar uses cookies or not but why it does what it does so Chrome report the use of cookies. And yes, local storage are also considered as cookies. Anything that stores data on the user's computer and can be used for tracking is considered a cookie. You cannot say that the laws on privacy and protection of personal data do not apply to you because you don't use cookies but local storage. So when Chrome report cookie usage and if you do not notify the end user that the website uses cookies (even though it doesn't), you may have legal problems.

**Reply by pdanpdan** [reply]
Citation needed for what you said.
Also can you please specify what cookie/local storage is set by default

**Comment by justinbeaty** [comment]
I was kind of curious about this, turns out the localStorage part at least is due to a line in `plugins/Platform.js`. Apparently, if you even access `window.localStorage` to see if it exists, Chrome now figures you're using localStorage.

Here's a sort of hacky fix for Quasar v1, probably would work for v2 as well.

In `quasar.conf.js`, make the following changes:

Import a few utils at the top of the file:

```
const fs = require('fs')
const path = require('path')

module.exports = function (/* ctx */) {
    // ...
```

Then, add the following under the `build` property:

```
extendWebpack(cfg, { isServer, isClient }) {
    // Define some paths                                                                                                                                                                                   
    const path_plugins_local = path.resolve(__dirname, 'src/plugins')
    const path_plugins_quasar = path.resolve(__dirname, 'node_modules/quasar/src/plugins')

    // Make a plugins folder, if not exist                                                                                                                                                                 
    fs.mkdirSync(path_plugins_local, { recursive: true })

    // Rewrite Platform module to remove reference to window.localStorage                                                                                                                                  
    const file_platform = fs.readFileSync(path.join(path_plugins_quasar, 'Platform.js'), 'utf8')
    fs.writeFileSync(path.join(path_plugins_local, 'Platform.js'), file_platform.replace('window.localStorage', 'false'))

    // Alias to our version
    cfg.resolve.alias[path.join(path_plugins_quasar, 'Platform.js')] = path.join(path_plugins_local, 'Platform.js')
},
```

The Quasar platform plugin will now report `false` for `Platform.has.webStorage`, but I suppose if you're not using it, it doesn't matter. Use at your own risk!

**Reply by rstoenescu** [reply]
@justinbeaty Pushed a commit to protect against evaluating window.localStorage until devland actually uses it. Will be available in Quasar v2.0.1.

**Reply by justinbeaty** [reply]
@rstoenescu Your solution of using a property getter is much better. Thanks!

**Comment by fjxhkj** [comment]
Hi guys, in Quasar v2.10.2, `$q.cookies.get()` returns an Object, why & how ?

```vue
setup() {
  const userLoginCookieName = 'userLogin'
  const $q = useQuasar()
  
  // set string of obj
  $q.cookies.set(userLoginCookieName, JSON.stringify({ name: '123456', pwd: '123123' }), {  expires: 3650  })

  if ($q.cookies.has(userLoginCookieName)) {
      const userLoginCookieInfo = $q.cookies.get(userLoginCookieName)
      console.log(typeof userLoginCookieInfo) // not string but an Object
  }
```
